### ISSUE TITLE
Flaky MismatchTest and MapMatcherTest tests

### ISSUE THREAD

**Issue by Apteryks** [mtrudel/machete#43]
Hi,

Using elixir-machete 0.3.3, running its test suite a few times, I can easily trigger these failures (with guix for example: `guix build --no-substitutes --no-grafts --rounds=20 elixir-machete`):

```
starting phase `check'
Running ExUnit with seed: 188457, max_cases: 192

..............................................................................................................................................

  1) test format_mismatches formats with numbers and indents (MismatchTest)
     test/machete/mismatch_test.exs:5
     Assertion with == failed
     code:  assert Machete.Mismatch.format_mismatches(mismatches, "  ") ==
              "  1) .a: 1 is not equal to 2\n  2) .b.[0].{0}: 1 is not equal to 2\n"
     left:  "  1) .b.[0].{0}: 1 is not equal to 2\n  2) .a: 1 is not equal to 2\n"
     right: "  1) .a: 1 is not equal to 2\n  2) .b.[0].{0}: 1 is not equal to 2\n"
     stacktrace:
       test/machete/mismatch_test.exs:8: (test)

.................................................................................................................................

  2) test produces a useful mismatch for max size mismatches (MapMatcherTest)
     test/machete/matchers/map_matcher_test.exs:29
     Assertion with ~> failed

     Mismatches:

       1) .[0].message: "%{b: 1, a: 1} is greater than 1 pairs in size" is not equal to "%{a: 1, b: 1} is greater than 1 pairs in size"

     code: assert %{a: 1, b: 1} ~>> map(max: 1) ~> mismatch("%{a: 1, b: 1} is greater than 1 pairs in size")
     stacktrace:
       test/machete/matchers/map_matcher_test.exs:30: (test)

..............................................................................................................................
Finished in 1.0 seconds (1.0s async, 0.01s sync)
204 doctests, 195 tests, 2 failures
error: in phase 'check': uncaught exception:
%exception #<&invoke-error program: "mix" arguments: ("do" "compile" "--no-deps-check" "--no-prune-code-paths" "+" "test" "--no-deps-check") exit-status: 2 term-signal: #f stop-signal: #f> 
phase `check' failed after 2.3 seconds
command "mix" "do" "compile" "--no-deps-check" "--no-prune-code-paths" "+" "test" "--no-deps-check" failed with status 2
build process 18 exited with status 256
builder for `/gnu/store/vbjswnk8ml05qxjvgi0l17k8mdvlbxpa-elixir-machete-0.3.3.drv' failed with exit code 1
```

**Comment by Apteryks** [comment]
The dependencies used are currently:  elixir-credo@1.7.7 elixir-dialyxir@1.4.3

**Comment by mtrudel** [comment]
I think this is a tooling problem on your side; I'm unable to reproduce it via bare calls to `mix test` or even looping tests within a custom mix task

**Comment by mtrudel** [comment]
There's nothing remotely process related in machete, so I don't see where the flakiness could be coming from an internal race.

**Comment by Apteryks** [comment]
Would you like to give it a try using Guix? It's a package/environment manager that excels at reproducing build environments (thus you'd be able to see the bug and attempt to understand it on your side).

Given the test sometimes passes, I doubt there's an issue in the tooling on the Guix side.

**Comment by mtrudel** [comment]
I'm afraid I don't have the bandwidth to deal with bugs that aren't reproducible on baseline tooling; there are just too many variations for one person to keep track of.

In the interest of trying to get this reproducible on baseline:

* What version of Elixir / Erlang are you on? Which platform?
* Is it *always* those two tests that fail?
* What percentage (roughly) of test runs fail? How long should I have to try reproducing?

**Comment by Apteryks** [comment]
Elexir is at 1.17.3
Erlang is at 27.1.1
It's always roughly these two tests failing
Perhaps 5-10% of test runs fail.

I can still reproduce this easily here:
```
starting phase `check'
Running ExUnit with seed: 45864, max_cases: 64

...............................................................................................................................................................

  1) test format_mismatches formats with numbers and indents (MismatchTest)
     test/machete/mismatch_test.exs:5
     Assertion with == failed
     code:  assert Machete.Mismatch.format_mismatches(mismatches, "  ") ==
              "  1) .a: 1 is not equal to 2\n  2) .b.[0].{0}: 1 is not equal to 2\n"
     left:  "  1) .b.[0].{0}: 1 is not equal to 2\n  2) .a: 1 is not equal to 2\n"
     right: "  1) .a: 1 is not equal to 2\n  2) .b.[0].{0}: 1 is not equal to 2\n"
     stacktrace:
       test/machete/mismatch_test.exs:8: (test)

.....................................................................................

  2) test produces a useful mismatch for max size mismatches (MapMatcherTest)
     test/machete/matchers/map_matcher_test.exs:29
     Assertion with ~> failed

     Mismatches:

       1) .[0].message: "%{b: 1, a: 1} is greater than 1 pairs in size" is not equal to "%{a: 1, b: 1} is greater than 1 pairs in size"

     code: assert %{a: 1, b: 1} ~>> map(max: 1) ~> mismatch("%{a: 1, b: 1} is greater than 1 pairs in size")
     stacktrace:
       test/machete/matchers/map_matcher_test.exs:30: (test)

.........................................................................................................................................................
Finished in 0.1 seconds (0.1s async, 0.00s sync)
204 doctests, 195 tests, 2 failures
error: in phase 'check': uncaught exception:
%exception #<&invoke-error program: "mix" arguments: ("do" "compile" "--no-deps-check" "--no-prune-code-paths" "+" "test" "--no-deps-check") exit-status: 2 term-signal: #f stop-signal: #f> 
phase `check' failed after 0.4 seconds
command "mix" "do" "compile" "--no-deps-check" "--no-prune-code-paths" "+" "test" "--no-deps-check" failed with status 2
build process 18 exited with status 256
builder for `/gnu/store/cq4mz66bb0a5fj21l7s74pk58mj5cc6w-elixir-machete-0.3.3.drv' failed with exit code 1
la compilation de /gnu/store/cq4mz66bb0a5fj21l7s74pk58mj5cc6w-elixir-machete-0.3.3.drv a échoué
Vous trouverez le journal de compilation dans « /var/log/guix/drvs/cq/4mz66bb0a5fj21l7s74pk58mj5cc6w-elixir-machete-0.3.3.drv ».
guix build: erreur : build of `/gnu/store/cq4mz66bb0a5fj21l7s74pk58mj5cc6w-elixir-machete-0.3.3.drv' failed
```

The package graph of `elixir-machete` in Guix, with the versions shown:
```
$ guix graph elixir-machete | dot -Tsvg -ograph.svg
```
![graph](https://github.com/user-attachments/assets/eb63799d-12d3-4845-a37a-5e2e865faf2e)

**Comment by mtrudel** [comment]
I can repro it locally. It's a version regression (likely due to Erlang 27; there's been a number of changes to how maps are implemented internally that induce these sorts of ordering errors). I'll get it fixed directly!

**Comment by mtrudel** [comment]
Fixed on main, released as `0.3.4`. Thanks again for the report!

**Comment by Apteryks** [comment]
That's a good outcome, thank you for persevering!
